src/commands.rs

```diff
@@ -734,9 +734,100 @@ implement_commands! {
     }
 }
 
+/// Allows pubsub callbacks to stop receiving messages.
+///
+/// Arbitraty data may be returned from `Break`.
+pub enum ControlFlow<U> {
+    Continue,
+    Break(U),
+}
+
+/// The PubSub trait allows subscribing to one or more items
+/// and receiving a callback whenever a message arrives.
+///
+/// Each method handles subscribing to the list of keys, waiting for
+/// messages, and unsubscribing from the same list of keys once
+/// a ControlFlow::Break is encountered.
+///
+/// Once (p)subscribe returns Ok(U), the connection is again safe to use
+/// for calling other methods.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// # fn do_something() -> redis::RedisResult<()> {
+/// use redis::{PubSubCommands, ControlFlow};
+/// let client = redis::Client::open("redis://127.0.0.1/")?;
+/// let mut con = client.get_connection()?;
+/// let mut count = 0;
+/// con.subscribe(&["foo"], |msg| {
+///     // do something with message
+///     assert_eq!(msg.get_channel(), Ok(String::from("foo")));
+///
+///     // incerement messages seen counter
+///     count += 1;
+///     match count {
+///         // stop after receiving 10 messages
+///         10 => ControlFlow::Break(()),
+///         _ => ControlFlow::Continue,
+///     }
+/// });
+/// # Ok(()) }
+/// ```
+//
+// In the future, it would be nice to implement Try such that `?` will work
+// within the closure.
+pub trait PubSubCommands {
+    /// Subscribe to a list of channels using SUBSCRIBE and run the provided
+    /// closure for each message received.
+    fn subscribe<'a, C, F, U>(&mut self, _: C, _: F) -> RedisResult<U>
```

Comment:
It might be good to explicitly state here that this is blocking until `Break` is returned. What do you think?

Reply:
Good idea, I'll add it.